crumbles/src/lib.rs

```diff
@@ -0,0 +1,653 @@
+                        io::ErrorKind::InvalidInput,
+                        "File too large for memory region",
+                    ));
+                }
+
+                let ptr = libc::mmap(
+                    inner.bytes.as_mut_ptr().cast(),
+                    file_len,
+                    PROT_READ,
+                    MAP_PRIVATE | MAP_FIXED,
+                    file.into_raw_fd(),
+                    offset as _,
+                );
+
+                if ptr == MAP_FAILED {
+                    return Err(io::Error::last_os_error());
+                }
+            }
+        }
+
+        Ok(inner)
+    }
+
+    unsafe fn set_dirty(&mut self, page_num: usize) -> io::Result<()> {
+        let start_addr = self.bytes.as_ptr() as usize;
+
+        let page_offset = page_num * PAGE_SIZE;
+        let page_addr = start_addr + page_offset;
+
+        if libc::mprotect(page_addr as _, PAGE_SIZE, PROT_READ | PROT_WRITE)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        if !self.last_snapshot().contains_key(&page_num) {
+            let mut clean_page = vec![0; PAGE_SIZE];
+            clean_page.copy_from_slice(&self.bytes[page_offset..][..PAGE_SIZE]);
+
+            self.last_snapshot_mut().insert(page_num, clean_page);
+        }
+
+        Ok(())
+    }
+
+    unsafe fn snap(&mut self) -> io::Result<()> {
+        if libc::mprotect(self.bytes.as_mut_ptr().cast(), PAGE_SIZE, PROT_READ)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        self.snapshots.push(Snapshot::new());
+
+        Ok(())
+    }
+
+    unsafe fn apply(&mut self) -> io::Result<()> {
+        if libc::mprotect(self.bytes.as_mut_ptr().cast(), PAGE_SIZE, PROT_READ)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        let popped_snapshot = self.snapshots.pop().unwrap();
+        if self.snapshots.is_empty() {
+            self.snapshots.push(Snapshot::new());
+        }
+        let snapshot = self.last_snapshot_mut();
+
+        for (page_num, clean_page) in popped_snapshot {
+            snapshot.entry(page_num).or_insert(clean_page);
+        }
+
+        Ok(())
+    }
+
+    unsafe fn revert(&mut self) -> io::Result<()> {
+        let popped_snapshot = self.snapshots.pop().unwrap();
+        if self.snapshots.is_empty() {
+            self.snapshots.push(Snapshot::new());
+        }
+
+        for (page_num, clean_page) in popped_snapshot {
+            let page_offset = page_num * PAGE_SIZE;
+            self.bytes[page_offset..][..PAGE_SIZE]
+                .copy_from_slice(&clean_page[..]);
+        }
+
+        if libc::mprotect(self.bytes.as_mut_ptr().cast(), PAGE_SIZE, PROT_READ)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(())
+    }
+
+    fn last_snapshot(&self) -> &Snapshot {
+        self.snapshots.last().unwrap()
```

Comment:
potentially adding a comment as to why the unwrap is safe (there is always at least one snapshot present)